### DISCUSSION TITLE
Requesting help with time analysis of quantum circuits

### DISCUSSION THREAD

**Discussion by FarooqKhaleel** [qutip/qutip-qip#69] [Q&A] [answered]
![20210630_160903](https://user-images.githubusercontent.com/86245057/123966390-0f619880-d9be-11eb-866c-544c897262d6.jpg)
Dear Respected Members of the group,
I hope you are doing well.
1) I kindly ask about how could I generate the time diagram for the quantum circuits as shown in the CNOT gate in the attachment?
2) How could I change the frequency rate of these pulses ?
Thank you very much for your help.
Yours sincerely,

**Comment by BoxiLi** [comment]
Sorry, but I don't really get the question you are studying. What is the y axis in your plot, some expectation value or the control pulses? If they are pulses, why are `y` and `O/P2` different? They are the same qubit.

**Comment by FarooqKhaleel** [comment]
Thank you for your reply and your help dear @BoxiLi
|y> is changing from |0> to |1> to |0> and so on and it has a phase shift of 90 degrees with respect to |x>. 
|x> is similar to |o/p1> they are interconnected with a quantum wire.

The above signals generate all the possibilities of the states that could enter the CNOT gate and, consequently, it generates the following
I/P..............O/P
|00> ........... |00>
|01> ........... |01>
|10> ........... |11>
|11> ........... |10>

I want to make every qubit ((|x>) and (|y>)) a serial clock of |0> and |1> that enter the CNOT and observe |o/p1> and |o/p2> as a time-dependent pulses as shown in the figure that I drew earlier.

Thank you very much for your help @BoxiLi 
Yours sincerely,

**Reply by BoxiLi** [reply]
Ok... I'm not sure if I get it correctly.

It seems that your diagram is not about the quantum circuit process at all but just about the relation between input and output. What the circuit does is that you give it |x> and |y> and it gives you the output. And you are neglecting the operation time of CNOT, so |o/p1> and |o/p2> change immediately after |x> and |y> change. How long does |x> stay in 0 is entire unto you.

If that is true, you just generate a time sequence, define a corresponding input sequence as you wish and compute and plot the output sequence, using `matplotlib` or whatever tools you prefer.

**Comment by FarooqKhaleel** [comment]
Dear @BoxiLi 
Thank you for your reply.
As a matter of fact, I want to give the CNOT gate pulses of |0> and |1> and the circuit compute the output accordingly.
It does not depend on CNOT gate only. I want the qutip-qip (not me) compute the output for larger circuits. However, I took the CNOT as an example.
The pulses I draw in the first comment (with a phase shift of 90) provide the total input combination to the circuit. (|00>, |01>, |10>, and |11>). If larger circuits are available, it will time-consuming that I compute the output and just draw it with matplotlib. 
Matplotlib just draws the output pulses, it does not compute the output. 
The goal is to make qutip-qip compute the output for several cascaded quantum gates and draw the timing chart for them. 
I wonder if there is such possibility?
Thank you for your time, efforts and your informative replies.
Yours sincerely,

**Comment by FarooqKhaleel** [comment]
The quantum gate's rise time and fall time in addition to the quantum wire's delays and possible noise sources are also important in the computation. I hope if they can also be included.
However, I just need qutip-qip draw the chart that I draw in the first comment. and I will deal with the rest.
Thank you very much.

**Comment by BoxiLi** [comment, marked as answer]
Ok, now I roughly get it.

As far as I understand, you want to give the circuit different inputs and get different outputs. `qutip-qip` don't have this build-in. The method implemented in `qutip-qip` only takes one input and gives you one output. So one run at a time. However, if your circuit does not change, then I feel like a `for` loop should be able to solve your question.

You could define a sequence of input states and run the circuit for each of them
``` python
time_sequence = [0, 1, 2, ,3 ,4, 5 ...]
input_state_sequence = [state0, state1,state2, ...]
output_state_sequence = []
for input_state in input_state_sequence :
    output = circuit.run(input_state)
    output_state_sequence .append(output)
```
And then, as you said, `matplotlib` can help you plot the figures. Basically, what `qutip-qip` can help is computing the circuit result. The rest will need to be handled with Python programming or other packages.

If you want to generate all possible input states on the computational basis automatically, I recommend using Python `itertools.product`
https://docs.python.org/3/library/itertools.html#itertools.product

**Comment by FarooqKhaleel** [comment]
Dear @BoxiLi 
Thank you very much for your detailed answer. 
I will try to write the script that generate the timing diagram using the hint you propose. I think it will achieve the purpose.
Thank you very much for your time and efforts and all the best with your study.
Yours sincerely,
